### ISSUE TITLE
'RNN' KeyError

### ISSUE THREAD

**Issue by jramapuram** [NVIDIA/apex#36]
Note that this is with the latest commit  12dce88d30cde716cc5ad1ea1af3e618a7c540e7

```python
In [9]: from apex import amp
   ...: amp_handle = amp.init()
   ...:                
   ...:                   
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-9-ef5bcdad1b52> in <module>()
      1 from apex import amp    
----> 2 amp_handle = amp.init()                                                   
                      
/opt/conda/lib/python3.6/site-packages/apex-0.1-py3.6.egg/apex/amp/amp.py in init(enabled, enable_caching, verbose, allow_banned)
    144                                         
    145     # 5.5) Extra-special handling of RNN backend              
--> 146     wrap.rnn_cast(torch.nn.backends.thnn.backend, 'RNN', verbose)
    147                                
    148     # And even more special handling of `backward` for fused gru / lstm
                                      
/opt/conda/lib/python3.6/site-packages/apex-0.1-py3.6.egg/apex/amp/wrap.py in rnn_cast(backend, fn, verbose)
    142 #   2) Insert an fp16 `flat_weight` if necessary
    143 def rnn_cast(backend, fn, verbose=False):        
--> 144     orig_rnn = utils.get_func(backend, fn)    
    145     @functools.wraps(orig_rnn)                                                                                                                                                                               
    146     def rnn_wrapper(*args, **kwargs):               

/opt/conda/lib/python3.6/site-packages/apex-0.1-py3.6.egg/apex/amp/utils.py in get_func(mod, fn)
    117 def get_func(mod, fn):       
    118     if isinstance(mod, torch.nn.backends.backend.FunctionBackend):
--> 119         return mod.function_classes[fn]
    120     else:             
    121         return getattr(mod, fn)             
                                     
KeyError: 'RNN'
```

**Comment by mcarilli** [comment]
This might have something to do with Pytorch's recent refactor of their RNN backend:  https://github.com/pytorch/pytorch/pull/10481

@cbcase, can you check if Amp works with current top of tree pytorch?  It appears the python exposure of their rnn backend has also changed:  https://github.com/pytorch/pytorch/pull/10481/files#diff-8f17185b332ae560499da6d86b99a2afR732

**Comment by mcarilli** [comment]
@jramapuram In the meantime, if you wish to use Amp, you may have to use a Pytorch version prior to 10481.

**Comment by cbcase** [comment]
Yup, those changes will mean refactoring how amp handles rnns and rnn cells -- I'll take a look through what's involved, but I expect it will be a substantial change and may take a little time.

Until then, any release version of PyTorch (ie, 0.4) should work but compiling from PyTorch master won't

**Comment by jramapuram** [comment]
@mcarilli  : I tried compiling pytorch using the commit before the one you mentioned but it still doesn't work. Guess only <0.5 works. Using apex 034b8f029c77f4ad4781b7f0a7f7ff31cb0e258d compiled against pytorch `484395e`

```bash
(base) ➜  ~ ipython
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [3]: from apex import amp

In [4]: amp_handle = amp.init()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-18310afd031d> in <module>()
----> 1 amp_handle = amp.init()

~/.venv3/lib/python3.6/site-packages/apex/amp/amp.py in init(enabled, enable_caching, verbose, allow_banned)
    151     # TODO: where else is this a problem?
    152     for rnn_type in ['GRUFused', 'LSTMFused']:
--> 153         mod = getattr(torch.nn._functions.thnn.rnnFusedPointwise, rnn_type)
    154         wrap.disable_casts(mod, 'backward', handle)
    155 

AttributeError: module 'torch.nn._functions.thnn' has no attribute 'rnnFusedPointwise'

In [5]: import torch

In [6]: print(torch.__version__)
0.5.0a0+484395e

```

**Comment by mcarilli** [comment]
It seems their RNN refactor actually began earlier:  https://github.com/pytorch/pytorch/commit/be5fb8f6fdb7dc21610401fef10113910690d3d1

Try using the parent of that one, e221791.  Sorry for the inconvenience, we're still working to make Amp compatible with the new changes.

**Comment by jinserk** [comment]
Hi @mcarilli, 
I met such RNN key error, so downgraded PyTorch to `e221791`, but still have some problems as:
Attempt to access StorageImpl having data type Half as data type Float (data at /u3/setup/pytorch/pytorch/aten/src/ATen/StorageImpl.h:74)
frame #0: THCudaTensor_data + 0x2c (0x7f7c8e20d2ec in /home/jbaik/.pyenv/versions/3.7.0/lib/python3.7/site-packages/torch/lib/libcaffe2_gpu.so)
frame #1: gpu_ctc + 0x37 (0x7f7cdd5fb127 in /home/jbaik/.pyenv/versions/3.7.0/lib/python3.7/site-packages/warpctc_pytorch-0.1-py3.7-linux-x86_64.egg/warpctc_pytorch/_warp_ctc/__warp_ctc.cpython-37m-x86_64-linux-gnu.so)
frame #2: <unknown function> + 0x1a1f (0x7f7cdd5faa1f in /home/jbaik/.pyenv/versions/3.7.0/lib/python3.7/site-packages/warpctc_pytorch-0.1-py3.7-linux-x86_64.egg/warpctc_pytorch/_warp_ctc/__warp_ctc.cpython-37m-x86_64-linux-gnu.so)
<omitting python frames>
frame #4: THPModule_safeCall(_object*, _object*, _object*) + 0x4c (0x7f7d5f79e13c in /home/jbaik/.pyenv/versions/3.7.0/lib/python3.7/site-packages/torch/_C.cpython-37m-x86_64-linux-gnu.so)
frame #12: THPFunction_apply(_object*, _object*) + 0x658 (0x7f7d5f98fca8 in /home/jbaik/.pyenv/versions/3.7.0/lib/python3.7/site-packages/torch/_C.cpython-37m-x86_64-linux-gnu.so)
frame #40: __libc_start_main + 0xf5 (0x7f7d7c7dc445 in /lib64/libc.so.6)
frame #41: python() [0x40069e]

does this work with [warp-ctc](https://github.com/SeanNaren/warp-ctc) pytorch binding?

**Comment by cbcase** [comment]
Fixed in #41 -- @jramapuram, please give the latest code a try!
